engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/MetadataStorage.java

```diff
@@ -0,0 +1,214 @@
+package org.enso.compiler.core.ir;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+import org.enso.compiler.core.CompilerStub;
+
+import scala.Option;
+
+/** Stores metadata for the various passes.
+  */
+public final class MetadataStorage {
+  private Map<ProcessingPass, ProcessingPass.Metadata> metadata;
+
+  public MetadataStorage() {
+    this(Collections.emptyMap());
+  }
+
+  public MetadataStorage(Map<ProcessingPass, ProcessingPass.Metadata> init) {
+    this.metadata = init;
+  }
+
+  /** Adds a new metadata entity to the pass metadata, or updates it if it
+    * already exists for a given pass.
+    *
+    * @param pass the pass to add the metadata for
+    * @param newMeta the metadata to add for `pass`
+    * @tparam K the concrete type of `pass`
+    */
+  public void update(ProcessingPass pass, ProcessingPass.Metadata newMeta) {
+    var copy = copyMetaMap();
```

Comment:
Why do we need to copy the map every time?

Reply:
That's what the original Scala implementation was doing as well - create new `Map` and reassign it to `metadata`.

Reply:
What's actually interesting is to compare:
- performance
- (multi threaded) consistency

@hubertp  managed to get [lazy immutable Scala map](https://github.com/enso-org/enso/pull/8359/commits/f9c34a94dca334411f9565eca4cac9ef3fb5778d) working. With that we can choose - do we want to use `scala...immutable.Map` here or immutable `java.util.Map` here? It is a valid question and I haven't yet chosen my answer.